File: Minesweeper/cpp/Minesweeper.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <random>
#include <queue>
#include <utility>

using namespace std;

int input_linenumber();
int input_columnnumber();
int input_bombnumber();

int know_digit(int); //桁を知る。

void display_board(vector<vector<char>> &board, int linenumber, int columnnumber);

void initialize_board(vector<vector<char>> &display_board_vector, vector<vector<int>> &data_board_vector, int linenumber, int columnnumber, int bombnumber);
void count_around_bomb(vector<vector<int>> &data_board_vector, int linenumber, int columnnumber);

void find_blank_space(vector<vector<char>> &display_board_vector, vector<vector<int>> &data_board_vector, int line_dig_point, int column_dig_point);

bool judge_game_clear(vector<vector<char>> &display_board_vector, vector<vector<int>> &data_board_vector, int linenumber, int columnnumber);

int main()
{
    const int linenumber = input_linenumber();
    const int columnnumber = input_columnnumber();
    const int bombnumber = input_bombnumber();

    int spare_bomb_number = bombnumber;
    
    vector<vector<char>> display_board_vector(linenumber,vector<char>(columnnumber,'o'));
    //旗はF、空は あるのは数字、まだ掘ってないのはoにする。
    vector<vector<int>> data_board_vector(linenumber,vector<int>(columnnumber,0));

    display_board(display_board_vector, linenumber, columnnumber);
    initialize_board(display_board_vector, data_board_vector, linenumber, columnnumber, bombnumber);

    char command;
    int line_dig_point,column_dig_point;
    while(true){
        if(judge_game_clear(display_board_vector, data_board_vector, linenumber,columnnumber)){
            display_board(display_board_vector, linenumber, columnnumber);
            cout << "Conguratulation! Clear!" << endl;
            break;
        }
        display_board(display_board_vector, linenumber, columnnumber);
        cout << "sparebombnumber:";
        cout << spare_bomb_number << endl;
        while(true){
            cout << "d:dig, f:bulid a flag, c:cancel a flag" << endl;
            cout << "input command" << endl;
            cin >> command;
            if(command == 'd'){
                cout << "where to dig" << endl;
                break;
            }else if(command == 'f'){
                cout << "where to build a flag" << endl;
                break;
            }else if(command == 'c'){
                cout << "which flag cancel" << endl;
            }else{
                cout << "input is incorrect" << endl;
            }
        }
        while(true){
            line_dig_point = input_linenumber();
            --line_dig_point;
            column_dig_point = input_columnnumber();
            --column_dig_point;
            try{
                if(display_board_vector.at(line_dig_point).at(column_dig_point)){
                    break;
                }
            }catch(exception &e){
                cout << "input is incorrect" << endl;
                continue;
            }
        }
        if(command == 'd'){
            if(display_board_vector.at(line_dig_point).at(column_dig_point) == 'F'){
                cout << "there is a flag!" << endl;
                continue;
            }else if(display_board_vector.at(line_dig_point).at(column_dig_point) != 'o'){
                cout << "there can't dig" << endl;
                continue;
            }else if(data_board_vector.at(line_dig_point).at(column_dig_point) == -1){
                display_board_vector.at(line_dig_point).at(column_dig_point) = 'x';
                display_board(display_board_vector, linenumber, columnnumber);
                cout << "there is bomb..." << endl;
                cout << "Game Over" << endl;
                break;
            }else if(data_board_vector.at(line_dig_point).at(column_dig_point) == 0){
                find_blank_space(display_board_vector, data_board_vector, line_dig_point, column_dig_point);
            }else{
                display_board_vector.at(line_dig_point).at(column_dig_point) = '0' + data_board_vector.at(line_dig_point).at(column_dig_point);
            }
        }else if(command == 'f'){
            if(display_board_vector.at(line_dig_point).at(column_dig_point) != 'o'){
                cout << "there can't build a flag" << endl;
                continue;
            }else{
                display_board_vector.at(line_dig_point).at(column_dig_point) = 'F';
                --spare_bomb_number;
            }
        }else if(command == 'c'){
            if(display_board_vector.at(line_dig_point).at(column_dig_point) != 'F'){
                cout << "there isn't a flag" << endl;
                continue;
            }else{
                display_board_vector.at(line_dig_point).at(column_dig_point) = 'o';
                ++spare_bomb_number;
            }
        }else{
            cout << "input is incorrect" << endl;
            continue;
        }
    }
    return 0;
}

int input_linenumber()
{
    int linenumber;
    cout << "input number of line" << endl;
    cin >> linenumber;
    return linenumber;
}

int input_columnnumber()
{
    int columnnumber;
    cout << "input number of column" << endl;
    cin >> columnnumber;
    return columnnumber;
}

int input_bombnumber()
{
    int bombnumber;
    cout << "input number of bomn" << endl;
    cin >> bombnumber;
    return bombnumber;
}

void display_board(vector<vector<char>> &board, int linenumber, int columnnumber)
{
    const int line_digit = know_digit(linenumber);
    const int column_digit = know_digit(columnnumber);
    int line,column,i;
    for(line = 0;line <= linenumber;++line){
        if(line == 0){
            for(i = 0;i < line_digit;++i){
                cout << ' ';
            }
            for(column = 1;column <= columnnumber;++column){
                cout << column;
                for(i = 0;i < column_digit - know_digit(column);++i){
                    cout << ' ';
                }
            }
            cout << endl;
        }else{
            cout << line;
            for(i = 0;i < line_digit - know_digit(line);++i){
                cout << ' ';
            }
            for(column = 1;column <= columnnumber;++column){
                cout << board.at(line-1).at(column-1);
                for(i = 0;i < column_digit - 1;++i){
                    cout << ' ';
                }
            }
            cout << endl;
        }
    }
}

void initialize_board(vector<vector<char>> &display_board_vector, vector<vector<int>> &data_board_vector, int linenumber, int columnnumber, int bombnumber)
{
    int line_dig_point,column_dig_point;
    cout << "input first dig point" << endl;
    while(true){
        line_dig_point = input_linenumber();
        --line_dig_point;
        column_dig_point = input_columnnumber();
        --column_dig_point;
        try{
            if(display_board_vector.at(line_dig_point).at(column_dig_point)){
                break;
            }
        }catch(exception &e){
            cout << "input is incorrect" << endl;
            continue;
        }
    }
    display_board_vector.at(line_dig_point).at(column_dig_point) = ' ';
    //爆弾を-1で表すことにする。
    vector<int> decide_where_is_bomb;
    int i,j;
    vector<int> around_the_first_dig_point;
    for(i = -1;i <=1;++i){
        for(j = -1;j <= 1;++j){
            try{
                if(data_board_vector.at(line_dig_point+i).at(column_dig_point+j) == 0){
                    around_the_first_dig_point.push_back((line_dig_point+i)*columnnumber+column_dig_point+j);
                }
            }catch(exception& e){
                continue;
            }
        }
    }
    for(i = 0;i <= linenumber*columnnumber-1;++i){
        if(binary_search(around_the_first_dig_point.begin(),around_the_first_dig_point.end(),i)){
            continue;
        }else{
            decide_where_is_bomb.push_back(i);
        }
    }
    random_device seed_gen;
    mt19937 engine(seed_gen());
    shuffle(decide_where_is_bomb.begin(), decide_where_is_bomb.end(), engine);
    for(i = 0;i < bombnumber;++i){
        data_board_vector.at(decide_where_is_bomb.at(i)/columnnumber).at(decide_where_is_bomb.at(i)%columnnumber) = -1;
    }
    count_around_bomb(data_board_vector, linenumber, columnnumber);
    find_blank_space(display_board_vector, data_board_vector, line_dig_point, column_dig_point);
}

void count_around_bomb(vector<vector<int>> &data_board_vector, int linenumber, int columnnumber)
{
    int line,column;
    int i,j;
    for(line = 0;line < linenumber;++line){
        for(column = 0;column < columnnumber;++column){
            if(data_board_vector.at(line).at(column) == -1){
                continue;
            }else{
                for(i = -1;i <=1;++i){
                    for(j = -1;j <= 1;++j){
                        if(i == 0 && j == 0){
                            continue;
                        }else{
                            try{
                                if(data_board_vector.at(line+i).at(column+j) == -1){
                                    ++data_board_vector.at(line).at(column);
                                }
                            }catch(exception& e){
                                continue;
                            }
                        }
                    }
                }
            } 
        }
    }
}

void find_blank_space(vector<vector<char>> &display_board_vector, vector<vector<int>> &data_board_vector, int line_dig_point, int column_dig_point)
{
    queue<pair<int,int>> find_blank;
    int line,column;
    find_blank.emplace(line_dig_point,column_dig_point);
    int i,j;
    while(!find_blank.empty()){
        line = find_blank.front().first;
        column = find_blank.front().second;
        for(i = -1;i <= 1;++i){
            for(j = -1;j <= 1;++j){
                try{
                    if(display_board_vector.at(line+i).at(column+j) == 'o'){
                        if(data_board_vector.at(line+i).at(column+j) == 0){
                            display_board_vector.at(line+i).at(column+j) = ' ';
                            find_blank.emplace(line+i,column+j);
                        }else{
                            display_board_vector.at(line+i).at(column+j) = '0' + data_board_vector.at(line+i).at(column+j);
                        }
                    }else{
                        continue;
                    }
                }catch(exception &e){
                    continue;
                }
            }   
        }
        find_blank.pop();
    }
}

bool judge_game_clear(vector<vector<char>> &display_board_vector, vector<vector<int>> &data_board_vector, int linenumber, int columnnumber)
{
    int line,column;
    for(line = 0;line < linenumber;++line){
        for(column = 0;column < columnnumber;++column){
            if(data_board_vector.at(line).at(column) != -1){
                if(display_board_vector.at(line).at(column) == 'o' || display_board_vector.at(line).at(column) == 'F'){
                    return false;
                }else{
                    continue;
                }
            }else{
                continue;
            }
        }
    }
    return true;
}

int know_digit(int number)
{
    int digit = 0;
    while(number){
        number /= 10;
        ++digit;
    }
    return digit;
}

File: minesweeper/Python/minesweeper.py
import numpy as np
import copy
import random
import queue

def first_input_linenumber():
    while True:
        print("縦のマスの数を入れてください。大きすぎると上手く表示出来なくなります。")
        temp = input("")
        if temp.isdecimal() and temp[0] != "0":
            return int(temp)
        else:
            print("自然数を入れてください")
            continue

def first_input_columnnumber():
    while True:
        print("横のマスの数を入れてください。大きすぎると上手く表示出来なくなります。")
        temp = input("")
        if temp.isdecimal() and temp[0] != "0":
            return int(temp)
        else:
            print("自然数を入れてください")
            continue
def first_input_bombnumber(linenumber,columnnumber):
    while True:
        print("地雷の数を入れてください")
        temp = input("")
        if temp.isdecimal() and temp[0] != "0":
            if int(temp) >= linenumber*columnnumber-8:
                print("地雷の数が多すぎます")
                continue
            else:
                return int(temp)
        else:
            print("自然数を入れてください")
            continue

linenumber = first_input_linenumber()
columnnumber = first_input_columnnumber()
bombnumber = first_input_bombnumber(linenumber,columnnumber)
display_board_array = np.full((linenumber,columnnumber),"o")
data_board_array = np.zeros((linenumber,columnnumber))

def know_digit(number):
    digit = 0
    while number:
        number //= 10
        digit += 1
    return digit

def display_board_func(display_board_array,linenumber,columnnumber):
    line_digit = know_digit(linenumber)
    column_digit = know_digit(columnnumber)
    for line in range(linenumber+1):
        if line == 0:
            for i in range(line_digit):
                print(" ",end="")
            for column in range(1,columnnumber+1):
                print(column,end="")
                for i in range(column_digit-know_digit(column)):
                    print(" ",end="")
            print()
        else:
            print(line,end="")
            for i in range(line_digit-know_digit(line)):
                print(" ",end="")
            for column in range(1,columnnumber+1):
                print(display_board_array[line-1][column-1],end="")
                for i in range(column_digit-1):
                    print(" ",end="")
            print()

display_board_func(display_board_array,linenumber,columnnumber)

def input_linenumber(linenumber):
    while True:
        print("マスの位置の縦の座標を入れてください")
        temp = input("")
        if temp.isdecimal() and temp[0] != "0":
            if int(temp) <= linenumber:
                return int(temp)
            else:
                print("そのマスはありません")
                continue
        else:
            print("自然数を入れてください")
            continue

def input_columnnumber(columnnumber):
    while True:
        print("マスの位置の横の座標を入れてください")
        temp = input("")
        if temp.isdecimal() and temp[0] != "0":
            if int(temp) <= columnnumber:
                return int(temp)
            else:
                print("そのマスはありません")
                continue
        else:
            print("自然数を入れてください")
            continue

def count_around_bomb(data_board_array,linenumber,columnnumber):
    threenumber = [-1,0,1]
    for line in range(linenumber):
        for column in range(columnnumber):
            if data_board_array[line][column] == -1:
                continue
            else:
                for i in threenumber:
                    for j in threenumber:
                        if i == 0 and j == 0:
                            continue
                        elif line+i < 0 or column+j < 0:
                            continue
                        else:
                            try:
                                if data_board_array[line+i][column+j] == -1:
                                    data_board_array[line][column] += 1
                            except IndexError:
                                continue

def find_blank_space(display_board_array,data_board_array,line_dig_point,column_dig_point):
    find_blank = queue.Queue()
    find_blank.put([line_dig_point,column_dig_point])
    while not find_blank.empty():
        temp = find_blank.get()
        line = temp[0]
        column = temp[1]
        threenumber = [-1,0,1]
        for i in threenumber:
            for j in threenumber:
                if line+i < 0 or column+j < 0:
                    continue
                else:
                    try:
                        if display_board_array[line+i][column+j] == "o":
                            if data_board_array[line+i][column+j] == 0:
                                display_board_array[line+i][column+j] = " "
                                find_blank.put([line+i,column+j])
                            else:
                                display_board_array[line+i][column+j] = str(data_board_array[line+i][column+j])
                        else:
                            continue
                    except IndexError:
                        continue


def initialize_board_func(display_board_array,data_board_array,linenumber,columnnumber,bombnumber):
    print("最初に掘るマスの場所を入れてください")
    line_dig_point = input_linenumber(linenumber)
    line_dig_point -= 1
    column_dig_point = input_columnnumber(columnnumber)
    column_dig_point -= 1
    display_board_array[line_dig_point][column_dig_point] = " "
    decide_where_is_bomb = []
    around_the_first_dig_point = []
    threenumber = [-1,0,1]
    for i in threenumber:
        for j in threenumber:
            if line_dig_point+i < 0 or column_dig_point+j < 0:
                continue
            try:
                if data_board_array[line_dig_point+i][column_dig_point+j] == 0:
                    around_the_first_dig_point.append((line_dig_point+i)*columnnumber+column_dig_point+j)
            except IndexError:
                continue
    for i in range(linenumber*columnnumber):
        if i in around_the_first_dig_point:
            continue
        else:
            decide_where_is_bomb.append(i)
    random.shuffle(decide_where_is_bomb)
    for i in range(bombnumber):
        data_board_array[decide_where_is_bomb[i]//columnnumber][decide_where_is_bomb[i]%columnnumber] = -1
    count_around_bomb(data_board_array,linenumber,columnnumber)
    find_blank_space(display_board_array,data_board_array,line_dig_point,column_dig_point)

initialize_board_func(display_board_array,data_board_array,linenumber,columnnumber,bombnumber)

def judge_game_clear(display_board_array,data_board_array,linenumber,columnnumber):
    for line in range(linenumber):
        for column in range(columnnumber):
            if data_board_array[line][column] != -1:
                if display_board_array[line][column] == "o" or display_board_array[line][column] == "F":
                    return False
                else:
                    continue
            else:
                continue
    return True

def display_bomb(display_board_array,data_board_array,linenumber,columnnumber):
    for line in range(linenumber):
        for column in range(columnnumber):
            if data_board_array[line][column] == -1:
                display_board_array[line][column] = "x"
            else:
                continue

sparebombnumber = copy.deepcopy(bombnumber)
while True:
    if judge_game_clear(display_board_array,data_board_array,linenumber,columnnumber):
        display_board_func(display_board_array,linenumber,columnnumber)
        print("おめでとうございます！ゲームクリアです！")
        break
    display_board_func(display_board_array,linenumber,columnnumber)
    print("残りの地雷数:",end="")
    print(sparebombnumber)
    while True:
        print("d:掘る, f;旗を立てる, c:旗を消す")
        command = input("行動を入力してください:")
        if command == "d":
            print("どこを掘りますか?")
            break
        elif command == "f":
            print("どこに旗を建てますか?")
            break
        elif command == "c":
            print("どこの旗を消しますか?")
            break
        else:
            print("入力が間違っています")
            continue
    line_dig_point = input_linenumber(linenumber)
    line_dig_point -= 1
    column_dig_point = input_columnnumber(columnnumber)
    column_dig_point -= 1
    if command == "d":
        if display_board_array[line_dig_point][column_dig_point] == "F":
            print("旗があります！")
            continue
        elif display_board_array[line_dig_point][column_dig_point] != "o":
            print("そこは掘れません")
            continue
        elif data_board_array[line_dig_point][column_dig_point] == -1:
            display_board_array[line_dig_point][column_dig_point] = "x"
            display_board_func(display_board_array,linenumber,columnnumber)
            print("残念！地雷がありました...")
            display_bomb(display_board_array,data_board_array,linenumber,columnnumber)
            display_board_func(display_board_array,linenumber,columnnumber)
            print("GAME OVER")
            break
        elif data_board_array[line_dig_point][column_dig_point] == 0:
            find_blank_space(display_board_array,data_board_array,line_dig_point,column_dig_point)
        else:
            display_board_array[line_dig_point][column_dig_point] = str(data_board_array[line_dig_point][column_dig_point])
    elif command == "f":
        if display_board_array[line_dig_point][column_dig_point] != "o":
            print("そこに旗は立てられません")
            continue
        else:
            display_board_array[line_dig_point][column_dig_point] = "F"
            sparebombnumber -= 1
    elif command == "c":
        if display_board_array[line_dig_point][column_dig_point] != "F":
            print("そこに旗はありません")
            continue
        else:
            display_board_array[line_dig_point][column_dig_point] = "o"
            sparebombnumber += 1
    else:
        print("入力が間違っています")